protocol/x/clob/keeper/deleveraging.go

```diff
@@ -188,6 +188,56 @@ func (k Keeper) CanDeleverageSubaccount(
 	return false, clobPair.Status == types.ClobPair_STATUS_FINAL_SETTLEMENT, nil
 }
 
+// GateWithdrawalsIfNegativeTncSubaccountSeen gates withdrawals if a negative TNC subaccount exists.
+// It does this by inserting a zero-fill deleveraging operation into the operations queue iff any of
+// the provided negative TNC subaccounts are still negative TNC.
+func (k Keeper) GateWithdrawalsIfNegativeTncSubaccountSeen(
+	ctx sdk.Context,
+	negativeTncSubaccountIds []satypes.SubaccountId,
+) (err error) {
+	foundNegativeTncSubaccount := false
+	var negativeTncSubaccountId satypes.SubaccountId
+	for _, subaccountId := range negativeTncSubaccountIds {
+		bigNetCollateral,
+			_,
+			_,
+			err := k.subaccountsKeeper.GetNetCollateralAndMarginRequirements(
+			ctx,
+			satypes.Update{SubaccountId: subaccountId},
+		)
+		if err != nil {
+			return err
+		}
+
+		// If the subaccount has negative TNC, mark that a negative TNC subaccount was found.
+		if bigNetCollateral.Sign() == -1 {
+			foundNegativeTncSubaccount = true
+			negativeTncSubaccountId = subaccountId
+			break
+		}
+	}
+
+	if !foundNegativeTncSubaccount {
+		return nil
+	}
+
+	// A negative TNC subaccount was found, therefore insert a zero-fill deleveraging operation into
+	// the operations queue to indicate withdrawals should be gated.
+	subaccount := k.subaccountsKeeper.GetSubaccount(ctx, negativeTncSubaccountId)
+	perpetualPositions := subaccount.GetPerpetualPositions()
+	if len(perpetualPositions) == 0 {
+		return errorsmod.Wrapf(
+			types.ErrNoPerpetualPositionsToLiquidate,
+			"GateWithdrawalsIfNegativeTncSubaccountSeen: subaccount has no open positions: (%s)",
+			lib.MaybeGetJsonString(subaccount),
+		)
+	}
+	perpetualId := subaccount.PerpetualPositions[0].PerpetualId
+	k.MemClob.InsertZeroFillDeleveragingIntoOperationsQueue(ctx, negativeTncSubaccountId, perpetualId)
+
+	return nil
```

Comment:
honestly would prefer to avoid this. the current approach is simpler since it avoids modifying existing critical paths to add on a hack (this is prob not long-term withdrawal gating solution)

for example if we modify `memclob.DeleverageSubaccount`, then wouldn't we have to update the logic to handle cases where zero-fills are returned from [`OffsetSubaccountPerpetualPosition`](https://github.com/dydxprotocol/v4-chain/blob/6014ec261478d509e1ffa6742b40fa5967e970a9/protocol/x/clob/keeper/deleveraging.go#L223)? maybe it's fine to add these to the op queue, but imo just makes things more complicated for little gain

given this is not the long-term solution, would rather add this new function so it's easy to rip out later once we build long-term solution for withdrawal gating

Reply:
yeah sounds good